### PR TITLE
NAS-123042 / 24.04 / Relax restrictions on NFS subdir sharing

### PR DESCRIPTION
**Pull request by mgrimesix**
What is restricted on sharing to the same path:
1) Repeated 'host' or 'network' names
2) Potential collisions with the use of wildcards (everyone) and
   overlaping of network subnets

### CODE REVIEW

File: src/middlewared/middlewared/plugins/nfs.py [hunk 1]
```diff
@@ -400,14 +403,28 @@ async def validate(self, data, schema_name, verrors, old=None):
...
         dns_cache = await self.resolve_hostnames(
-            sum([share["hosts"] for share in other_shares], []) + data["hosts"]
+            sum([share["hosts"] for share in other_shares], []) + data_hosts
         )
+
+        self.validate_share_networks_input(data['networks'], dns_cache, schema_name, verrors)
+
+        # Stop here if the input generated errors for the user to fix
+        if verrors:
+            return
```
Comment (anodos325) [inline src/middlewared/middlewared/plugins/nfs.py:419]: You could probably just verrors.check()

File: src/middlewared/middlewared/plugins/nfs.py [hunk 2]
```diff
@@ -442,6 +459,84 @@ async def validate(self, data, schema_name, verrors, old=None):
...
+        The host field can contain multiple 'host' entries that are space delimited
+        The validation is limited to detecting repeated 'host' names
+
+        This will return a list of hosts
+        """
+        new_host_list = []
+        for host_str in hosts:
+            # Trap space delimited hosts
+            if ' ' in host_str:
+                new_host_list.extend(shlex.split(host_str))
```
Comment (anodos325) [inline src/middlewared/middlewared/plugins/nfs.py:473]: We should raise a ValidationError if users are trying to cram multiple addresses /  networks as a space-delimited list.

File: src/middlewared/middlewared/plugins/nfs.py [hunk 3]
```diff
@@ -442,6 +459,84 @@ async def validate(self, data, schema_name, verrors, old=None):
...
+            dups = list(set([x for x in networks if networks.count(x) >= 2]))
+            verrors.add(
+                f"{schema_name}.networks",
+                f"ERROR - Duplicate network entries are not allowed: {dups}"
+            )
+        # Check for duplicates via hostname (which should be in dns_cache)
+        for IPaddr in networks:
+            IPinterface = ipaddress.ip_interface(IPaddr)
+
+            if str(IPinterface.ip) in list(dns_cache.values()):
```
Comment (yocalebo) [inline src/middlewared/middlewared/plugins/nfs.py:516]: We're generating a list of the dns_cache values for each ip specified. We should generate this information outside the for loop just once.

```
dns_cache_values = list(dns_cache.values())
for IPaddr in networks:
    IPinterface = ipaddress.ip_interface(IPaddr)
    if str(IPinterface.ip) in dns_cache_values:
        ...

File: src/middlewared/middlewared/plugins/nfs.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mgrimesix) [reply src/middlewared/middlewared/plugins/nfs.py:473]: After talking to @anodos325 I'm going to block multiple host entries in the single field. So, no spaces or quotes in this field

File: src/middlewared/middlewared/plugins/nfs.py [hunk 4]
```diff
@@ -528,74 +646,105 @@ def validate_hosts_and_networks(self, other_shares, data, schema_name, verrors,
...
+            /mnt/tank/dataset 192.168.1.1(rw)
+            /mnt/tank/dataset *(ro)
+        2) A host could be part of a network subnet, e.g.
+            /mnt/tank/dataset 192.168.1.1(rw)
+            /mnt/tank/dataset 192.168.1.0/24(ro)
+            Note: The 192.168.1.0/24 is a 'networks' specification
+        This function checks for common conditions.
         """
+        # We test other shares that are sharing the same path
+        tgt_path_stat = os.stat(data["path"])
```
[the remark this answers is not in the document]
Comment (anodos325) [reply src/middlewared/middlewared/plugins/nfs.py:704]: To be more specific. If you need to stat you can call `filesystem.stat` via our middleware API.

File: src/middlewared/middlewared/plugins/nfs.py [hunk 5]
```diff
@@ -442,6 +459,84 @@ async def validate(self, data, schema_name, verrors, old=None):
...
+            )
+        # Check for duplicates via hostname (which should be in dns_cache)
+        for IPaddr in networks:
+            IPinterface = ipaddress.ip_interface(IPaddr)
+
+            if str(IPinterface.ip) in list(dns_cache.values()):
+                key = next(key for key, value in dns_cache.items() if value == str(IPinterface.ip))
+                verrors.add(
+                    f"{schema_name}.networks",
+                    f"ERROR - Duplicate network entries are not allowed: {key} {IPaddr}"
```
[the remark this answers is not in the document]
Comment (mgrimesix) [reply src/middlewared/middlewared/plugins/nfs.py:520]: yep, you are correct.

File: src/middlewared/middlewared/plugins/nfs.py [hunk 6]
```diff
@@ -528,74 +607,105 @@ def validate_hosts_and_networks(self, other_shares, data, schema_name, verrors,
...
-                lowerpath = share['path']
-                upperpath = data['path']
-                found_match = True
-
-            if found_match:
+            try:
+                share_stat = await self.middleware.call('filesystem.stat', share["path"])
+                share_dev = share_stat['dev']
+                share_ino = share_stat['inode']
+            except Exception:
```
[the remark this answers is not in the document]
Comment (anodos325) [reply src/middlewared/middlewared/plugins/nfs.py:675]:
```
except CallError as e:
    if e.errno != errno.ENOENT:
        self.logger.warning(<message>)
        
   continue
```

File: tests/api2/test_300_nfs.py [hunk 7]
```diff
@@ -16,6 +16,7 @@
 from functions import PUT, POST, GET, SSH_TEST, DELETE, wait_on_job
 from functions import make_ws_request
 from auto_config import pool_name, ha, hostname
+# from auto_config import interface
```
Comment (anodos325) [inline tests/api2/test_300_nfs.py:19]: Should this be removed?

**Review by yocalebo — APPROVED** [review]
I'll approve on condition that Andrew's latest comments are addressed. Thanks!

**Comment by mgrimesix** [comment]
After convo with Andrew, a small bit of refactoring and add check for symlinks

**Review by anodos325 — APPROVED** [review]

**Pull request merged**